pkg/block/mem/adapter.go

```diff
@@ -359,20 +451,140 @@ func (a *Adapter) ResolveNamespace(_, storageNamespace, key string, identifierTy
+	a.mutex.RLock()
+	defer a.mutex.RUnlock()
+	storageID := obj.StorageID
+	if a.mpu[storageID] == nil {
+		return nil, ErrMultiPartNotFound
+	}
+	mpu, ok := a.mpu[storageID][uploadID]
+	if !ok {
+		return nil, ErrMultiPartNotFound
+	}
+	// Collect part numbers and sort
+	partNumbers := make([]int, 0, len(mpu.parts))
+	for pn := range mpu.parts {
+		partNumbers = append(partNumbers, pn)
+	}
+	sort.Ints(partNumbers)
+	// Pagination
+	startIdx := 0
+	if opts.PartNumberMarker != nil {
+		marker, err := strconv.Atoi(*opts.PartNumberMarker)
+		if err != nil {
+			return nil, fmt.Errorf("invalid part number marker: %w", err)
+		}
+		for i, pn := range partNumbers {
+			if pn > marker {
+				startIdx = i
+				break
+			}
+		}
+	}
+	maxParts := len(partNumbers)
+	if opts.MaxParts != nil && int(*opts.MaxParts) < maxParts-startIdx {
+		maxParts = startIdx + int(*opts.MaxParts)
+	}
+	isTruncated := false
+	var nextPartNumberMarker *string
+	if maxParts < len(partNumbers) {
+		isTruncated = true
+		nextMarker := strconv.Itoa(partNumbers[maxParts-1])
+		nextPartNumberMarker = &nextMarker
+	}
+	parts := make([]block.MultipartPart, 0, maxParts-startIdx)
+	for _, pn := range partNumbers[startIdx:maxParts] {
+		p := mpu.parts[pn]
+		etag := calcETag(p.data)
+		parts = append(parts, block.MultipartPart{
+			ETag:         etag,
+			PartNumber:   pn,
+			LastModified: p.lastModified,
+			Size:         p.size,
+		})
+	}
+	return &block.ListPartsResponse{
+		Parts:                parts,
+		NextPartNumberMarker: nextPartNumberMarker,
+		IsTruncated:          isTruncated,
+	}, nil
 }
-func (a *Adapter) ListMultipartUploads(_ context.Context, _ block.ObjectPointer, _ block.ListMultipartUploadsOpts) (*block.ListMultipartUploadsResponse, error) {
-	return nil, block.ErrOperationNotSupported
+
+func (a *Adapter) ListMultipartUploads(_ context.Context, obj block.ObjectPointer, opts block.ListMultipartUploadsOpts) (*block.ListMultipartUploadsResponse, error) {
+	a.mutex.RLock()
+	defer a.mutex.RUnlock()
+	storageID := obj.StorageID
+	type mpuEntry struct {
+		id  string
+		mpu *mpu
+	}
+	mpuList := make([]mpuEntry, 0)
+	if a.mpu[storageID] != nil {
+		for id, m := range a.mpu[storageID] {
+			mpuList = append(mpuList, mpuEntry{id, m})
+		}
+	}
+	// Sort by key, then upload ID (lexicographically)
+	sort.Slice(mpuList, func(i, j int) bool {
+		if mpuList[i].mpu.objectKey == mpuList[j].mpu.objectKey {
+			return mpuList[i].id < mpuList[j].id
+		}
+		return mpuList[i].mpu.objectKey < mpuList[j].mpu.objectKey
+	})
+	// Pagination
+	startIdx := 0
+	if opts.KeyMarker != nil && *opts.KeyMarker != "" {
+		for i, entry := range mpuList {
+			if entry.mpu.objectKey > *opts.KeyMarker {
+				startIdx = i
+				break
+			}
+		}
+	}
+	if opts.UploadIDMarker != nil && *opts.UploadIDMarker != "" {
+		for i, entry := range mpuList[startIdx:] {
+			if entry.id > *opts.UploadIDMarker {
+				startIdx += i
```

Comment:
This seems very strange!  Are we sure this is correct?  (Does the adapter test test listing well enough to be sure?)

Reply:
Tested manually:
- Found a bug https://github.com/treeverse/lakeFS/issues/9035
- Fixed the code to align the s3 behaviour and address UploadIDMarker only if KeyMarker is set.